File: frontend/src/data/postos.ts
/** Postos e graduações do Exército Brasileiro - Civil a Gen Ex */
export interface PostoGraduacao {
  id: number;
  nome: string;
  abrev: string;
}

export const POSTOS: PostoGraduacao[] = [
  { id: 6, nome: "Terceiro Sargento", abrev: "3º Sgt" },
  { id: 7, nome: "Segundo Sargento", abrev: "2º Sgt" },
  { id: 8, nome: "Primeiro Sargento", abrev: "1º Sgt" },
  { id: 9, nome: "Subtenente", abrev: "STen" },
  { id: 10, nome: "Aspirante", abrev: "Asp" },
  { id: 11, nome: "Segundo Tenente", abrev: "2º Ten" },
  { id: 12, nome: "Primeiro Tenente", abrev: "1º Ten" },
  { id: 13, nome: "Capitão", abrev: "Cap" },
  { id: 14, nome: "Major", abrev: "Maj" },
  { id: 15, nome: "Tenente Coronel", abrev: "TC" },
  { id: 16, nome: "Coronel", abrev: "Cel" },
];

/** Mapa nome completo → abreviatura (cobre todos os valores do enum Python + legado) */
export const POSTO_ABREV: Record<string, string> = {
  ...Object.fromEntries(POSTOS.map((p) => [p.nome, p.abrev])),
  // Legado: registros criados antes da padronização para "Aspirante"
  "Aspirante a Oficial": "Asp Of",
};

/**
 * Formata referência ao usuário: "Abrev. Posto + Nome de Guerra" (ou nome completo como fallback).
 * Ex.: formatNomeComPosto("Gustavo Silva", "Capitão", "Silva") → "Cap Silva"
 */
export function formatNomeComPosto(
  nome: string,
  posto: string | null | undefined,
  nomeDeGuerra?: string | null,
): string {
  const displayName = nomeDeGuerra?.trim() || nome;
  if (!posto) return displayName;
  const abrev = POSTO_ABREV[posto];
  return abrev ? `${abrev} ${displayName}` : displayName;
}

File: backend/app/utils/postos.py
"""Abreviaturas de posto/graduação - espelha PostoGraduacaoEnum."""

POSTO_ABREV: dict[str, str] = {
    "Terceiro Sargento":    "3º Sgt",
    "Segundo Sargento":     "2º Sgt",
    "Primeiro Sargento":    "1º Sgt",
    "Subtenente":           "S Ten",
    "Aspirante a Oficial":  "Asp Of",
    "Segundo Tenente":      "2º Ten",
    "Primeiro Tenente":     "1º Ten",
    "Capitão":              "Cap",
    "Major":                "Maj",
    "Tenente-Coronel":      "Ten Cel",
    "Coronel":              "Cel",
    # Aliases para registros no banco criados antes da padronização
    "Aspirante":            "Asp Of",
    "Tenente Coronel":      "Ten Cel",
}


def abrev_posto(posto: str | None) -> str:
    """Converte valor do enum para abreviatura (ex.: 'Capitão' → 'Cap').
    Retorna string vazia se posto for None/vazio.
    Retorna o valor original se não houver mapeamento (fallback seguro).
    """
    if not posto:
        return ""
    return POSTO_ABREV.get(posto, posto)
